src/Ledger/TokenAlgebra/ValueRelation.lagda

```diff
@@ -0,0 +1,464 @@
+
+ ii : (Rv ⊕₁ Ru) ⇒ (Ru ⊕₁ Rv)
+ ii (qv , qu , Rqv , Rqu , q≡qvqu) = qu , qv , Rqu , Rqv , trans q≡qvqu (+-comm qv qu)
+
+
+⊕₁-assoc : Algebra.Associative _⇐⇒_ _⊕₁_
+⊕₁-assoc Ru Rv Rw = i , ii
+ where
+ i : ((Ru ⊕₁ Rv) ⊕₁ Rw) ⇒ (Ru ⊕₁ (Rv ⊕₁ Rw))
+ i {aid} {q} (quv , qw , (qu , qv , Ruqu , Rvqv , quv≡quqv) , Rwqw , q≡quvqw) = Goal
+  where
+  assc : q ≡ qu + (qv + qw)
+  assc = begin  q               ≡⟨ q≡quvqw ⟩
+                quv + qw        ≡⟨ cong (λ x → x + qw) quv≡quqv ⟩
+                qu + qv + qw    ≡⟨ +-assoc qu qv qw ⟩
+                qu + (qv + qw)  ∎
+
+  Goal : (Ru ⊕₁ (Rv ⊕₁ Rw)) aid q
+  Goal = qu , qv + qw , Ruqu , (qv , qw , Rvqv , Rwqw , refl) , assc
+
+ ii : (Ru ⊕₁ (Rv ⊕₁ Rw)) ⇒ ((Ru ⊕₁ Rv) ⊕₁ Rw)
+ ii {aid} {q} (qu , qvw , Ruqu , (qv , qw , Rvqv , Rwqw , qvw≡qvqw) , q≡quqvw) = Goal
+  where
+  assc : qu + qvw ≡ qu + qv + qw
+  assc = begin  qu + qvw        ≡⟨ cong (λ x → qu + x) qvw≡qvqw ⟩
+                qu + (qv + qw)  ≡⟨ sym (+-assoc qu qv qw) ⟩
+                qu + qv + qw    ∎
+
+  Goal : ((Ru ⊕₁ Rv) ⊕₁ Rw) aid q
+  Goal = qu + qv , qw , (qu , qv , Ruqu , Rvqv , refl) , Rwqw , trans q≡quqvw assc
+
+
+⊕-assoc : Algebra.Associative _≋_ _⊕_
+⊕-assoc (Ru , luu) (Rv , lvu) (Rw , lwu)  = i , ii
+ where
+ i : (Ru ⊕₁ Rv) ⊕₁ Rw ⇒ Ru ⊕₁ (Rv ⊕₁ Rw)
+ i {aid} {q} (quv , qw , (qu , qv , Ruqu , Rvqv , quv≡quqv) , Rwqw , q≡quvqw) = Goal
+  where
+  assc : quv + qw ≡ qu + (qv + qw)
+  assc = begin  quv + qw        ≡⟨ cong (λ x → x + qw) quv≡quqv ⟩
+                qu + qv + qw    ≡⟨ +-assoc qu qv qw ⟩
+                qu + (qv + qw)  ∎
+
+  Goal : (Ru ⊕₁ (Rv ⊕₁ Rw)) aid q
+  Goal = qu , ((qv + qw) , (Ruqu , ((qv , qw , Rvqv , Rwqw , refl) , trans q≡quvqw assc )))
+
+
+ ii : Ru ⊕₁ (Rv ⊕₁ Rw) ⇒ (Ru ⊕₁ Rv) ⊕₁ Rw
+ ii {aid} {q} (qu , qvw , Ruqu , (qv , qw , Rvqv , Rwqw , qvw≡qvqw) , q≡quqvw) = Goal
+  where
+  tras : q ≡ qu + qv + qw
+  tras = begin  q               ≡⟨ q≡quqvw ⟩
+                qu + qvw        ≡⟨ cong (λ x → qu + x) qvw≡qvqw ⟩
+                qu + (qv + qw)  ≡⟨ sym (+-assoc qu qv qw) ⟩
+                qu + qv + qw    ∎
+
+  Goal : ((Ru ⊕₁ Rv) ⊕₁ Rw) aid q
+  Goal = (qu + qv) , (qw , ((qu , qv , Ruqu , Rvqv , refl) , (Rwqw , tras)))
+
+
+⊕-cong : Algebra.Congruent₂ _≋_ _⊕_
+⊕-cong {Ru , luu} {Rv , lvu} {Ru' , luu'} {Rv' , lvu'} (Ru⇒Rv , Rv⇒Ru) (Ru'⇒Rv' , Rv'⇒Ru') = i , ii
+ where
+ i : (Ru ⊕₁ Ru') ⇒ (Rv ⊕₁ Rv')
+ i {aid} {q} (qu , qu' , Ruqu , Ru'qu' , q≡ququ') = qu , qu' , Ru⇒Rv Ruqu , Ru'⇒Rv' Ru'qu' , q≡ququ'
+
+ ii : (Rv ⊕₁ Rv') ⇒ (Ru ⊕₁ Ru')
+ ii {aid} {q} (qv , qv' , Rvqv , Rv'qv' , q≡qvqv') = qv , qv' , Rv⇒Ru Rvqv , Rv'⇒Ru' Rv'qv' , q≡qvqv'
+\end{code}
+
+\subsection{Definition of the value monoid}
+
+An inhabitant of `Value` is a map denoting a finite collection of quantities of assets.
+
+\begin{code}
+open TokenAlgebraPoly
+open CommutativeMonoid renaming (_∙_ to _⋆_) hiding (trans ; sym)
+open Algebra
+
+module _
+  {AdaPolicy   : PolicyID}
+  {DecEqValue  : DecidableEquality (AssetID ⇀ Quantity) }
```

Comment:
This is never inhabitable, even though I don't think that's provable. Think of the map that contains `1 Ada` if the Riemann hypothesis is true, and `0 Ada` otherwise. This is fixed by switching to a decidable set theory.

Reply:
I don't see why assuming a decidable set theory is better than assuming decidable equality here.  We won't ever construct an inhabitant/proof of `DecidableEquality (AssetID ⇀ Quantity)`.  That's why it's a module parameter; i.e., it's an assumption or "axiom."  In particular, it precludes having a map that contains 1 Ada if the Riemann hypothesis is actually a theorem and 0 if RH is false.

Reply:
It's an executable spec, and in particular we use that decidable equality in `consumed pp s tx ≡ produced pp s tx` in line 253 in `Ledger.Utxo` (or rather when we provide a `Computational` instance for that). Also, it's not true that this axiom precludes that. For example, here's a proof that all types have decidable equality using this axiom:
```
  Dec-all : {A : Set} → DecidableEquality A
  Dec-all {A} a b = Dmap
    (mk⇔ (λ h → helper (subst (λ M → proj₁ M ID 0) h (Prelude.refl , Prelude.refl , Prelude.refl)))
          (λ where refl → Prelude.refl)) (DecEqValue Amap Bmap)
    where open import Relation.Nullary.Decidable renaming (map to Dmap)
          ID : AssetID
          ID = AdaPolicy , AdaName

          genMap : A → AssetID ⇀ Quantity
          genMap X = (λ p x → p ≡ ID × x ≡ 0 × (a ≡ X))
                   , λ where (_ , refl , _) (_ , refl , _) → Prelude.refl

          Amap : AssetID ⇀ Quantity
          Amap = genMap a

          Bmap : AssetID ⇀ Quantity
          Bmap = genMap b

          helper : proj₁ Bmap ID 0 → a ≡ b
          helper (_ , _ , h) = h
```